### PR TITLE
[WIP] Rethink the PathMappingRepository

### PR DESCRIPTION
**Pull request by tgalopin**
The beta6 version of the PathMappingRepository was a beginning, but it's clear now that I didn't thought it well enough to take all the use-cases into account.

This PR aim to introduce a new, simpler and more efficient version of the PathMappingRepository to fix the issues https://github.com/puli/repository/pull/40 and https://github.com/puli/repository/pull/39. Once this PR will be ready, I'll also work on the windows problems (https://github.com/puli/issues/issues/102), the link resources (https://github.com/puli/issues/issues/107) and the versionning (https://github.com/puli/issues/issues/90).

### CODE REVIEW

**Comment by webmozart** [comment]
Thanks! pli/issues#107 and puli/issues#90 can wait for the next release. The other ones should be fixed in this release to fix the broken projects.

**Comment by tgalopin** [comment]
Then this PR should be the last step before beta7 :) .

**Comment by webmozart** [comment]
That's the goal, yes. :)

File: src/PathMappingRepository.php [hunk 1]
```diff
@@ -80,18 +81,54 @@ public function find($query, $language = 'glob')
...
-     * Not supported.
-     *
      * {@inheritdoc}
      */
     public function remove($query, $language = 'glob')
     {
-        throw new BadMethodCallException(sprintf('remove\(\) is not supported in %s', __CLASS__));
+        $query = $this->sanitizePath($query);
+
+        Assert::notEq('', trim($query, '/'), 'The root directory cannot be removed.');
```
Comment (webmozart) [inline src/PathMappingRepository.php]: I'd prefer if you used `notSame()` and `same()` whenever possible. However, here I'd use `notEmpty()`.

Comment (tgalopin) [inline src/PathMappingRepository.php]: We should remove it from where I copied it then :) : https://github.com/puli/repository/blob/1.0/src/FilesystemRepository.php#L238

Comment (webmozart) [inline src/PathMappingRepository.php]: Fine by me :)

Comment (tgalopin) [inline src/PathMappingRepository.php]: In this PR? I think we should create another PR to do that everywhere needed, it doesn't seem very related to this.

Comment (webmozart) [inline src/PathMappingRepository.php]: Yep let's do that.

**Comment by webmozart** [comment]
Reported to be working on Windows. :)

**Pull request merged**